tools/adev-api-extraction/index.mts

```diff
@@ -0,0 +1,129 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import {readFileSync, writeFileSync} from 'fs';
+import path from 'path';
+// @ts-ignore This compiles fine, but Webstorm doesn't like the ESM import in a CJS context.
+import {
+  ClassEntry,
+  CompilerOptions,
+  createCompilerHost,
+  DocEntry,
+  EntryCollection,
+  InterfaceEntry,
+  NgtscProgram,
+} from '@angular/compiler-cli';
+import ts from 'typescript';
+import {EXAMPLES_PATH, interpolateCodeExamples} from './interpolate-code-examples.mjs';
+
+function main() {
+  const [paramFilePath] = process.argv.slice(2);
+  const rawParamLines = readFileSync(paramFilePath, {encoding: 'utf8'}).split('\n');
+
+  const [
+    moduleName,
+    moduleLabel,
+    serializedPrivateModules,
+    entryPointExecRootRelativePath,
+    srcs,
+    outputFilenameExecRootRelativePath,
+    serializedPathMapWithExecRootRelativePaths,
+    extraEntriesSrcs,
+  ] = rawParamLines;
+
+  const privateModules = new Set(serializedPrivateModules.split(','));
+
+  // The path map is a serialized JSON map of import path to index.ts file.
+  // For example, {'@angular/core': 'path/to/some/index.ts'}
+  const pathMap = JSON.parse(serializedPathMapWithExecRootRelativePaths) as Record<string, string>;
+
+  // The tsconfig expects the path map in the form of path -> array of actual locations.
+  // We also resolve the exec root relative paths to absolute paths to disambiguate.
+  const resolvedPathMap: {[key: string]: string[]} = {};
+  for (const [importPath, filePath] of Object.entries(pathMap)) {
+    resolvedPathMap[importPath] = [path.resolve(filePath)];
+
+    // In addition to the exact import path,
+    // also add wildcard mappings for subdirectories.
+    const importPathWithWildcard = path.join(importPath, '*');
+    resolvedPathMap[importPathWithWildcard] = [
+      path.join(path.resolve(path.dirname(filePath)), '*'),
+    ];
+  }
+
+  const compilerOptions: CompilerOptions = {
+    paths: resolvedPathMap,
```

Comment:
(for the future): This is where we would read/incorporate the real tsconfig just FYI